File: classifier/data/raw/human/samples/project_euler_zig/problem_095.zig
const std = @import("std");

// Project Euler 95: smallest member of the longest amicable chain below one million.
pub fn main() !void {
    const limit: usize = 1_000_000;
    const allocator = std.heap.page_allocator;
    const sums = try allocator.alloc(usize, limit + 1);
    defer allocator.free(sums);
    @memset(sums, 0);
    for (1..limit / 2 + 1) |divisor| {
        var multiple = divisor * 2;
        while (multiple <= limit) : (multiple += divisor) sums[multiple] += divisor;
    }

    const completed = try allocator.alloc(bool, limit + 1);
    defer allocator.free(completed);
    @memset(completed, false);
    const seen_run = try allocator.alloc(u32, limit + 1);
    defer allocator.free(seen_run);
    @memset(seen_run, 0);
    const depth = try allocator.alloc(usize, limit + 1);
    defer allocator.free(depth);

    var run: u32 = 0;
    var best_length: usize = 0;
    var answer: usize = 0;
    for (1..limit + 1) |start| {
        if (completed[start]) continue;
        run += 1;
        var value = start;
        var length: usize = 0;
        while (value <= limit and !completed[value] and seen_run[value] != run) {
            seen_run[value] = run;
            depth[value] = length;
            length += 1;
            value = sums[value];
        }
        if (value <= limit and seen_run[value] == run) {
            const cycle_length = length - depth[value];
            if (cycle_length > best_length) {
                best_length = cycle_length;
                answer = value;
                var item = sums[value];
                while (item != value) : (item = sums[item]) answer = @min(answer, item);
            }
        }
        value = start;
        while (value <= limit and !completed[value] and seen_run[value] == run) {
            completed[value] = true;
            value = sums[value];
        }
    }
    var buffer: [32]u8 = undefined;
    try std.Io.File.writeStreamingAll(std.Io.File.stdout(), std.Io.Threaded.global_single_threaded.io(), (try std.fmt.bufPrint(&buffer, "{d}\n", .{answer})));
}

File: classifier/data/raw/ai/samples/project_euler_java/Problem095.java
// Project Euler 95: smallest member of the longest amicable chain below one million.
public class Problem095 {
    public static void main(String[] args) {
        final int limit = 1_000_000;
        int[] sum = new int[limit + 1];
        for (int d = 1; d <= limit / 2; d++) for (int k = d + d; k <= limit; k += d) sum[k] += d;
        boolean[] processed = new boolean[limit + 1];
        int[] visit = new int[limit + 1], path = new int[limit + 1];
        int bestLength = 0, bestMember = 0;
        for (int start = 2, run = 1; start <= limit; start++, run++) {
            if (processed[start]) continue;
            int x = start, size = 0;
            while (x <= limit && !processed[x] && visit[x] != run) {
                visit[x] = run; path[size++] = x; x = sum[x];
            }
            if (x <= limit && visit[x] == run) {
                int begin = 0; while (path[begin] != x) begin++;
                int length = size - begin, smallest = path[begin];
                for (int i = begin + 1; i < size; i++) smallest = Math.min(smallest, path[i]);
                if (length > bestLength) { bestLength = length; bestMember = smallest; }
            }
            for (int i = 0; i < size; i++) processed[path[i]] = true;
        }
        System.out.println(bestMember);
    }
}
